File: api/api_response.py
from enum import Enum
from typing import TypedDict, List, Optional, Dict, Any


class Status(Enum):
    OK = "ok"
    ERROR = "error"


class ResponseType(Enum):
    ERROR = "error"
    PHOTO_METRICS = "photo_metrics"


class SpeciesCount(TypedDict):
    animal: str
    count: int


class PhotoMetrics(TypedDict):
    file: str
    annotated_file: str
    predictions: List[SpeciesCount]


class ApiResponse:
    def __init__(
            self,
            status: Status,
            photo_metrics: Optional[List[PhotoMetrics]] = None,
            error: Optional[str] = None,
            session: Optional[str] = None
    ):
        self.status = status
        self.photo_metrics = photo_metrics
        self.error = error
        self.session = session

    def to_dict(self) -> Dict[str, Any]:
        return {
            'status': self.status.value,
            'photo_metrics': self.photo_metrics,
            'error': self.error,
            'session': self.session
        }

File: ui/src/ApiResponse.ts
import { useEffect, useState } from "react";

export type Status = "ok" | "error";

export interface SpeciesCount {
  animal: number;
  count: number;
}

export interface PhotoMetrics {
  file: string;
  predictions: Array<SpeciesCount>;
}

export interface PhotoMetricsRequest {
  files: Array<string>;
}

export interface ApiResponse {
  status: Status;
  photo_metrics: Array<PhotoMetrics> | undefined;
  error: string | undefined;
}

export function usePhotoMetrics(
  req: PhotoMetricsRequest
): ApiResponse | undefined {
  const [apiResponse, setApiResponse] = useState<ApiResponse | undefined>(
    undefined
  );
  useEffect(() => {
    fetchPhotoMetrics(req).then((data) => setApiResponse(data));
  }, [req.files.join(",")]);
  return apiResponse;
}

export function fetchPhotoMetrics(
  req: PhotoMetricsRequest
): Promise<ApiResponse> {
  return fetch("/api/v1/predict_counts", {
    method: "POST",
    headers: {
      Accept: "application/json",
      "Content-Type": "application/json",
    },
    body: JSON.stringify(req),
  })
    .then((resp) => resp.json())
    .then((data) => data as ApiResponse);
}